xcode/Subconscious/Shared/Parsers/Subtext.swift

```diff
@@ -651,33 +658,47 @@ extension Subtext {
 
 extension Subtext {
     private static let maxExcerptSize = 280
+    private static let maxExcerptBlocks = 2
     private static let maxGeneratedSlugSize = 128
     
     /// Derive an excerpt
-    func excerpt(fallback: String = "") -> String {
-        // Filter out empty blocks
-        let validBlocks = blocks
-            .filter { block in !block.isEmpty }
-            .map { block in String(block.span) }
-            .prefix(2) // Take first two blocks
+    func excerpt(fallback: String = "") -> Subtext {
+        let length = Self.maxExcerptSize * Self.maxExcerptBlocks
+        let blocks = self.blocks
+            .filter({ block in !block.isEmpty })
+            .prefix(Self.maxExcerptBlocks)
+        
+        // Fast path: if the content fits we just take the first `maxExcerptBlocks`
+        if self.toString().count < length {
+            return Subtext(base: base, blocks: Array(blocks))
+        }
         
-        let output = validBlocks.joined(separator: "\n")
+        // Otherwise we need to truncate the string
+        let text = String.truncateAtWordBoundary(markup: self.toString(), maxChars: length)
+        // Re-parse it
+        let dom = Subtext(markup: text)
         
-        return output.isEmpty ? fallback : output
+        // Then we can take the the first `maxExcerptBlocks` blocks
+        let excerptBlocks = dom.blocks
+            .filter({ block in !block.isEmpty })
+            .prefix(Self.maxExcerptBlocks)
+       
+        return Subtext(
+            base: dom.base,
+            blocks: Array(excerptBlocks)
+        )
     }
     
-    func truncate(_ maxBlocks: Int = 2, fallback: String = "") -> Subtext {
-        // Filter out empty blocks
-        let validBlocks = blocks
-            .filter { block in !block.isEmpty }
-            .prefix(maxBlocks)
-        
-        return Subtext(base: self.base, blocks: Array(validBlocks))
+    static func excerpt(markup: any StringProtocol, fallback: String = "") -> Subtext {
```

Comment:
My understanding is that `any` has a runtime cost vs generics. May want to write this as a generic with a `StringProtocol` constraint.

Reply:
Change to `<S: StringProtocol>` and tweaked the API, I believe this should avoid the `any` tax.